syntax = "proto3";

package order;

service OrderService {
    rpc CreateOrder(CreateOrderRequest) returns (OrderResponse);
    rpc GetOrder(GetOrderRequest) returns (OrderResponse);
    rpc ListOrders(ListOrdersRequest) returns (ListOrdersResponse);
    rpc UpdateOrderStatus(UpdateOrderStatusRequest) returns (OrderResponse);
    rpc CancelOrder(CancelOrderRequest) returns (OrderResponse);
}

message CreateOrderRequest {
    string user_id = 1;
    string cart_id = 2;
    string shipping_address = 3;
    string payment_method = 4;
    repeated OrderItem items = 5;
}

message OrderItem {
    string product_id = 1;
    string product_name = 2;
    float price = 3;
    int32 quantity = 4;
    float total = 5;
}

message GetOrderRequest {
    string order_id = 1;
}

message ListOrdersRequest {
    string user_id = 1;
    int32 page = 2;
    int32 limit = 3;
    string status = 4;
}

message UpdateOrderStatusRequest {
    string order_id = 1;
    string status = 2;
    string notes = 3;
}

message CancelOrderRequest {
    string order_id = 1;
    string reason = 2;
}

message OrderResponse {
    string order_id = 1;
    string user_id = 2;
    string cart_id = 3;
    repeated OrderItem items = 4;
    float subtotal = 5;
    float tax = 6;
    float shipping_cost = 7;
    float total_amount = 8;
    string status = 9; // PENDING, PROCESSING, SHIPPED, DELIVERED, CANCELLED
    string shipping_address = 10;
    string payment_method = 11;
    string created_at = 12;
    string updated_at = 13;
}

message ListOrdersResponse {
    repeated OrderResponse orders = 1;
    int32 total = 2;
    int32 page = 3;
    int32 limit = 4;
}